src/reflect/scala/reflect/internal/Variances.scala

```diff
@@ -205,14 +205,39 @@ trait Variances {
   }
 
   /** Compute variance of type parameter `tparam` in all types `tps`. */
-  def varianceInTypes(tps: List[Type])(tparam: Symbol): Variance =
-    fold(tps map (tp => varianceInType(tp)(tparam)))
+  def varianceInTypes(tps: List[Type])(tparam: Symbol): Variance = {
+    @tailrec def loop(ts: List[Type], acc: Variance): Variance =
+      if (ts.isEmpty || acc.isInvariant) acc
+      else loop(ts.tail, acc & varianceInType(ts.head)(tparam))
+    loop(tps, Bivariant)
+  }
 
   /** Compute variance of type parameter `tparam` in type `tp`. */
   def varianceInType(tp: Type)(tparam: Symbol): Variance = {
-    def inArgs(sym: Symbol, args: List[Type]): Variance = fold(map2(args, sym.typeParams)((a, p) => inType(a) * p.variance))
-    def inSyms(syms: List[Symbol]): Variance            = fold(syms map inSym)
-    def inTypes(tps: List[Type]): Variance              = fold(tps map inType)
+    def inArgs(sym: Symbol, args: List[Type]): Variance = {
+      @tailrec def loop(as: List[Type], bs: List[Symbol], acc: Variance): Variance =
+        if (acc.isInvariant || as.isEmpty && bs.isEmpty) acc
+        else loop(as.tail, bs.tail, acc & (inType(as.head) * bs.head.variance))
+      loop(args, sym.typeParams, Bivariant)
+    }
+    def inSyms(syms: List[Symbol]): Variance = {
+      @tailrec def loop(ss: List[Symbol], acc: Variance): Variance =
+        if (acc.isInvariant || ss.isEmpty) acc
+        else loop(ss.tail, acc & inSym(ss.head))
+      loop(syms, Bivariant)
+    }
+    def inTypes(tps: List[Type]): Variance = {
+      @tailrec def loop(ts: List[Type], acc: Variance): Variance =
+        if (acc.isInvariant || ts.isEmpty) acc
+        else loop(ts.tail, acc & inType(ts.head))
+      loop(tps, Bivariant)
+    }
+    def inAnnots(anns: List[AnnotationInfo]): Variance = {
+      @tailrec def loop(as: List[AnnotationInfo], acc: Variance): Variance =
+        if (acc.isInvariant || as.isEmpty) acc
+        else loop(as.tail, acc & inType(as.head.atp))
+      loop(anns, Bivariant)
+    }
```

Comment:
It would be possible to reduce the duplication without incurring value class boxing by leveraging our SAM support.

```
trait VarianceExtractor[T] { def apply(t: T): Variance }


def foldMap[T](ts: List[T])(extractor: VarianceExtractor[T]): Variance = { ... }

foldMap(syms)((sym: Symbol) => if (sym.isAliasType) inType(sym.info).cut else inType(sym.info)}
```

Reply:
Ok. Correct me if I am wrong: the boxing that the compiler cannot remove with a simple `Function1[-A, +B]`, it can remove it with the specific trait `VarianceExtractor`, _because_ any code that may be calling on the API of  `Function1` could not have the knowledge that `B` can be, in certain specific calls, a fixed type (`Variance`).

Reply:
Yes -- when you use instantiate a type parameter to value class type the boxed representation is used within the generic code, and box/unbox operations are added at the boundary between generic and concrete code.